gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JournalBasedTokenStateService.java

```diff
@@ -0,0 +1,190 @@
+                    super.addToken(id, issueTime, expiration, maxLifetime);
+
+                } catch (Exception e) {
+                    log.failedToLoadJournalEntry(id, e);
+                }
+            }
+        } catch (IOException e) {
+            throw new ServiceLifecycleException("Failed to load persisted state from the token state journal", e);
+        }
+    }
+
+    @Override
+    public void addToken(final String tokenId, long issueTime, long expiration, long maxLifetimeDuration) {
+        super.addToken(tokenId, issueTime, expiration, maxLifetimeDuration);
+
+        try {
+            journal.add(tokenId, issueTime, expiration, maxLifetimeDuration);
+        } catch (IOException e) {
+            log.failedToAddJournalEntry(tokenId, e);
+        }
+    }
+
+    @Override
+    public long getTokenExpiration(final String tokenId, boolean validate) throws UnknownTokenException {
+        // Check the in-memory collection first, to avoid file access when possible
+        try {
+            // If the token identifier is valid, and the associated state is available from the in-memory cache, then
+            // return the expiration from there.
+            return super.getTokenExpiration(tokenId, validate);
+        } catch (UnknownTokenException e) {
+            // It's not in memory
+        }
+
+        if (validate) {
+            validateToken(tokenId);
+        }
+
+        // If there is no associated state in the in-memory cache, proceed to check the journal
+        long expiration = 0;
+        try {
+            JournalEntry entry = journal.get(tokenId);
+            if (entry == null) {
+                throw new UnknownTokenException(tokenId);
+            }
+
+            expiration = Long.parseLong(entry.getExpiration());
+            super.addToken(tokenId,
+                           Long.parseLong(entry.getIssueTime()),
+                           expiration,
+                           Long.parseLong(entry.getMaxLifetime()));
+        } catch (IOException e) {
+            log.failedToLoadJournalEntry(e);
+        }
+
+        return expiration;
+    }
+
+    @Override
+    protected long getMaxLifetime(final String tokenId) {
+        long result = super.getMaxLifetime(tokenId);
+
+        // If there is no result from the in-memory collection, proceed to check the journal
+        if (result < 1L) {
+            try {
+                JournalEntry entry = journal.get(tokenId);
+                if (entry == null) {
+                    throw new UnknownTokenException(tokenId);
+                }
+                result = Long.parseLong(entry.getMaxLifetime());
+                super.setMaxLifetime(tokenId, Long.parseLong(entry.getIssueTime()), result);
+            } catch (Exception e) {
+                log.failedToLoadJournalEntry(e);
+            }
+        }
+        return result;
+    }
+
+    @Override
+    protected List<String> getTokens() {
+        List<String> tokenIds = new ArrayList<>();
+
```

Comment:
Should we be checking in-memory implementation first?

Reply:
Hmm, I think we _could_ since this implementation loads all the persisted token state on init(), and any subsequent addition/removal of state should be reflected in the cache.

Reply:
I think that would mean this override simply goes away though, rather than checking the superclass implementation _first_.